src/Swarm/Version.hs

```diff
@@ -0,0 +1,119 @@
+commitInfo = case gitInfo of
+  Left _ -> ""
+  Right git -> " (" <> giBranch git <> "@" <> take 10 (giHash git) <> ")"
+
+type Hash = String
+
+isSwarmReleaseTag :: String -> Bool
+isSwarmReleaseTag = all (\c -> isDigit c || c == '.')
+
+tagVersion :: Maybe (Hash, String)
+tagVersion = case gitInfo of
+  Left _ -> Nothing
+  Right gi ->
+    let t = giTag gi
+        ((ta, _num), ghash) = first (first init . breakOnEnd "-" . init) $ breakOnEnd "-" t
+     in if isSwarmReleaseTag ta
+          then Just (ghash, ta)
+          else Nothing
+
+version :: String
+version =
+  let v = showVersion Paths_swarm.version
+   in if v == "0.0.0.1" then "pre-alpha version" else v
+
+upstreamReleaseVersion :: IO (Maybe String)
+upstreamReleaseVersion = do
+  manager <- newManager tlsManagerSettings
+  -- -------------------------------------------------------------------------------
+  -- SEND REQUEST
+  request <- parseRequest "https://api.github.com/repos/swarm-game/swarm/releases"
+  response <-
+    httpLbs
+      request {requestHeaders = [(hUserAgent, "swarm-game/swarm-swarmversion")]}
+      manager
+  -- -------------------------------------------------------------------------------
+  -- PARSE RESPONSE
+  -- putStrLn $ "The status code was: " ++ show (statusCode $ responseStatus response)
+  let result = decodeEither' (BS.pack . BSL.unpack $ responseBody response) :: Either ParseException Array
+  case result of
+    Left _e -> do
+      -- print e
+      return Nothing
+    Right rs -> do
+      ts <- forM (toList rs) $ \r -> do
+        return . flip parseMaybe r $ \case
+          Object o -> o .: "tag_name"
+          _ -> fail "The JSON list does not contain structures!"
+      return $ find isSwarmReleaseTag . catMaybes $ ts
+
+data NewReleaseFailure where
+  NoUpstreamRelease :: NewReleaseFailure
+  OnDevelopmentBranch :: String -> NewReleaseFailure
+  OldUpstreamRelease :: String -> String -> NewReleaseFailure
+
+instance Show NewReleaseFailure where
+  show = \case
+    NoUpstreamRelease -> "No upstream releases found."
+    OnDevelopmentBranch br -> "Currently on development branch '" <> br <> "', skipping release query."
+    OldUpstreamRelease upTag myTag -> "Upstream release '" <> upTag <> "' is not newer than mine ('" <> myTag <> "')."
+
+getNewerReleaseVersion :: IO (Either NewReleaseFailure String)
+getNewerReleaseVersion =
+  case gitInfo of
+    -- when using cabal install, the git info is unavailable, which is of no interest to players
+    Left _e -> maybe (Left NoUpstreamRelease) Right <$> upstreamReleaseVersion
+    Right gi ->
+      if giBranch gi /= "main"
+        then return . Left . OnDevelopmentBranch $ giBranch gi
+        else do
+          mUpTag <- upstreamReleaseVersion
+          return $ case mUpTag of
+            Nothing -> Left NoUpstreamRelease
+            Just upTag -> case snd <$> tagVersion of
+              Nothing -> Right upTag
+              Just myTag ->
+                if myTag >= upTag
```

Comment:
Heh, it would work for some time and then break on the 10th release. :smile: 

There is `Data.Version` in `base` which is very simple and used in autogenerated `Paths_swarm.version`.
From the docs, it looks like it will cover our needs. :slightly_smiling_face:

Reply:
Btw. `versions` is a hilarious library. As long as we stick to PVP in Cabal and in tags, we should not need it.

I might filter Github prereleases from the list just to be safe though, we might want to use those in the future.